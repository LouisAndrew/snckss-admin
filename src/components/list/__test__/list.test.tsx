import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

import { render, cleanup, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'

import List, { ListItem } from '..'

describe('List reusable Component', () => {
        let concatenated: ListItem['key'] = []

        const mockHandleRemove = jest.fn((key: ListItem['key']) => {
                concatenated = [...concatenated, key]
        })

        const mockHandleClick = jest.fn((key: ListItem['key']) => {
                concatenated = [...concatenated, key]
        })

        const mockItems: ListItem[] = [
                {
                        text: 'item1',
                        key: 'item1',
                },
                {
                        text: 'item2',
                        key: 'item2',
                },
        ]

        const mockHeader: string = 'Mock Header'

        afterEach(cleanup)
        const el: React.ReactElement = (
                <List
                        items={mockItems}
                        headerText={mockHeader}
                        handleRemove={mockHandleRemove}
                        handleClick={mockHandleClick}
                />
        )

        it('renders without crashing', () => {
                const div = document.createElement('div')
                ReactDOM.render(el, div)
        })

        it('renders with children correctly', () => {
                const { getByText } = render(el)

                // iterate over mockItems to make sure it exists within the document
                mockItems.forEach((item) => {
                        const listItem: HTMLElement = getByText(item.text)
                        expect(listItem).toBeInTheDocument()
                })
        })

        it('renders and respond to user clicking the event correctly when clicked.', () => {
                const { getByText } = render(el)

                mockItems.forEach((item) => {
                        const el: HTMLElement = getByText(item.text)
                        fireEvent.click(el)

                        expect(concatenated).toContain(item.key)
                })
        })

        it('renders and removes the item correctly when an item is clicked', () => {
                const { getAllByTestId } = render(el)
                const listItem: HTMLElement[] = getAllByTestId('list-items')

                mockItems.forEach((item, index) => {
                        const deleteSpan: HTMLElement | null = document.querySelector(
                                `.${listItem[index].className
                                        .split(' ')
                                        .join('.')} .del-${item.text}`
                        )

                        if (deleteSpan) {
                                fireEvent.click(deleteSpan)
                                expect(concatenated).toContain(item.key)
                        } else {
                                fail()
                        }
                })
        })

        it('matches snapshot', () => {
                const tree = renderer.create(el).toJSON()
                expect(tree).toMatchSnapshot()
        })
})
