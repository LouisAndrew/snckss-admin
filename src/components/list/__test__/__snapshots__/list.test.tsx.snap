// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List reusable Component matches snapshot 1`] = `
<ul
  className="list-group"
>
  <label
    className="list-label"
  >
    Mock Header
  </label>
  <li
    className="list-group-item"
    data-testid="list-items"
  >
    <div
      className="text clickable"
      onClick={[Function]}
    >
      item1
    </div>
    <span
      className="del-item1"
      onClick={[Function]}
    >
      <svg
        dangerouslySetInnerHTML={
          Object {
            "__html": "<path d=\\"M19 4h-3.5l-1-1h-5l-1 1H5v2h14M6 19a2 2 0 0 0 2 2h8a2 2 0 0 0 2-2V7H6v12z\\" fill=\\"currentColor\\"/>",
          }
        }
        focusable={false}
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "transform": "rotate(360deg)",
            "verticalAlign": "-0.125em",
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      />
    </span>
  </li>
  <li
    className="list-group-item"
    data-testid="list-items"
  >
    <div
      className="text clickable"
      onClick={[Function]}
    >
      item2
    </div>
    <span
      className="del-item2"
      onClick={[Function]}
    >
      <svg
        dangerouslySetInnerHTML={
          Object {
            "__html": "<path d=\\"M19 4h-3.5l-1-1h-5l-1 1H5v2h14M6 19a2 2 0 0 0 2 2h8a2 2 0 0 0 2-2V7H6v12z\\" fill=\\"currentColor\\"/>",
          }
        }
        focusable={false}
        height="1em"
        preserveAspectRatio="xMidYMid meet"
        style={
          Object {
            "transform": "rotate(360deg)",
            "verticalAlign": "-0.125em",
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      />
    </span>
  </li>
</ul>
`;
